examples/pytorch/gat.py

```diff
@@ -0,0 +1,240 @@
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT License.
+"""Example of GAT model trained on Cora dataset."""
+
+from typing import List, Any, Tuple
+from dataclasses import dataclass
+import os.path as osp
+
+import numpy as np
+import ray
+import ray.train as train
+from ray.train.torch import TorchTrainer
+from ray.air import session
+from ray.air.config import ScalingConfig
+from sklearn.metrics import accuracy_score
+import torch
+from torch import nn
+import torch.nn.functional as F
+from torch.utils.data import DataLoader, IterableDataset
+from torch_geometric.nn import GATConv
+
+from deepgnn.graph_engine.data.citation import Cora
+from deepgnn.graph_engine import Graph, graph_ops
+from deepgnn.graph_engine.snark.distributed import Server, Client as DistributedClient
+
+
+@dataclass
+class GATQueryParameter:
+    """Configuration for graph query."""
+
+    neighbor_edge_types: np.ndarray
+    feature_idx: int
+    feature_dim: int
+    label_idx: int
+    label_dim: int
+    num_classes: int
+    batch_size: int = 70
+    feature_type: np.dtype = np.float32
+    label_type: np.dtype = np.float32
+    num_hops: int = 2
+
+
+class GATDataset(IterableDataset):
+    """Query to fetch graph data for the model."""
+
+    def __init__(self, query_param: GATQueryParameter, inputs: np.array, graph: Graph):
+        """Initialize graph query."""
+        super(GATDataset, self).__init__()
+        self.query_param = query_param
+        self.graph = graph
+        self.inputs = inputs
+        self.label_meta = np.array(
+            [[query_param.label_idx, query_param.label_dim]], np.int32
+        )
+        self.feat_meta = np.array(
+            [[query_param.feature_idx, query_param.feature_dim]], np.int32
+        )
+
+    def __iter__(self):
+        """
+        Create iterator over inputs.
+
+        We shuffle the inputs at the beginning of each epoch and drop last nodes not fitting
+        into a batch.
+        """
+        np.random.shuffle(self.inputs)
+        worker_info = torch.utils.data.get_worker_info()
+        batches = self.inputs
+        last_elements = len(self.inputs) % self.query_param.batch_size
+        if last_elements > 0:
+            batches = batches[:-last_elements]
+        batches = batches.reshape(-1, self.query_param.batch_size)
+
+        if worker_info is None:
+            return map(self.query, batches)
+
+        # split workload for distributed training
+        return map(self.query, batches[worker_info.id :: worker_info.num_workers])
+
+    def query(self, inputs: np.ndarray) -> tuple:
+        """Fetch training data."""
+        nodes, edges, src_idx = graph_ops.sub_graph(
```

Comment:
You should test this example on different datasets, last I remember this sub_graph method breaks on all but one dataset.

Reply:
This is an example for specific dataset, typically [other datasets have their own examples](https://github.com/pyg-team/pytorch_geometric/blob/master/examples/graph_sage_unsup_ppi.py).